compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2750,32 +2751,69 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                             | LifetimeBinderKind::PolyTrait
                             | LifetimeBinderKind::WhereBound
                     );
-                    let (span, sugg) = if span.is_empty() {
-                        let sugg = format!(
+
+                    let mut span = span;
+                    let mut rm_poly_trait_span = DUMMY_SP;
+                    let mut sugg = format!("{}, ", name.unwrap_or("'a"));
+
+                    if span.is_empty() {
+                        let mut generic_params = "".to_string();
+                        if let Some(with_poly_trait_ref) = self.with_poly_trait_ref.get(&node_id)
+                            && higher_ranked
+                        {
+                            generic_params = with_poly_trait_ref.generic_param_idents.iter().fold(
+                                "".to_string(),
+                                |mut generic_params, x| {
+                                    generic_params += x.as_str();
+                                    generic_params += ", ";
+                                    generic_params
+                                },
+                            );
+                            if !generic_params.is_empty() {
+                                rm_poly_trait_span = with_poly_trait_ref
+                                    .poly_trait
+                                    .with_hi(with_poly_trait_ref.trait_ref.lo());
+                            }
+                        }
+
+                        sugg = format!(
                             "{}<{}>{}",
                             if higher_ranked { "for" } else { "" },
-                            name.unwrap_or("'a"),
+                            format!("{}{}", generic_params, name.unwrap_or("'a")),
                             if higher_ranked { " " } else { "" },
                         );
-                        (span, sugg)
                     } else {
-                        let span = self
+                        span = self
                             .r
                             .tcx
                             .sess
                             .source_map()
                             .span_through_char(span, '<')
                             .shrink_to_hi();
-                        let sugg = format!("{}, ", name.unwrap_or("'a"));
-                        (span, sugg)
-                    };
+                    }
+
                     if higher_ranked {
```

Comment:
At the moment you add the suggestion if `higher_ranked` (etc.) but I *think* we should only do it if `kind` matches `WhereBound`. This might matter for code like:

```rs
trait Trait<T>
where
    T: B<'b> + for<'a> A<'a>,
{
}

trait A<'a> {}
trait B<'a> {}
```

Where we have 3 possible lifetime ribs to attach a lifetime to: `PolyTrait` followed by `WhereBound` followed by `Item`.

For reference, master emits the following 3 suggestions:

* `T: for<'b> B<'b> + for<'a> A<'a>,` OK (`PolyTrait`)
* `for<'b> T: B<'b> + for<'a> A<'a>,` BAD (nested quantifiers) (`WhereBound`)
* `trait Trait<'b, T>` OK (`Item`)

Could you check if your PR handles this code correctly? You likely only get a *single* suggestion (since you set `should_continue` to `false`), namely the `PolyTrait` one. I wonder if the `PolyTrait` one is still correct.